gui: Improve thread naming

hebasto opened the pull request:
On **master** (eef90c14ed0f559e3f6e187341009270b84f45cb):
- thread list from OS:
![Screenshot from 2020-04-28 00-25-41](https://user-images.githubusercontent.com/32963518/80425413-3de07100-88ec-11ea-8d7a-79bd9e152395.png)
- log excerpt:
```
2020-04-27T21:25:26Z [] GUI: initialize : Running initialization in thread
...
2020-04-27T21:26:04Z [] Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2020-04-27T21:26:04Z [] Using wallet /home/hebasto/.bitcoin/testnet3/wallets/test2
2020-04-27T21:26:04Z [] BerkeleyEnvironment::Open: LogDir=/home/hebasto/.bitcoin/testnet3/wallets/test2/database ErrorFile=/home/hebasto/.bitcoin/testnet3/wallets/test2/db.log
2020-04-27T21:26:04Z [] init message: Loading wallet...
2020-04-27T21:26:04Z [] BerkeleyEnvironment::Open: LogDir=/home/hebasto/.bitcoin/testnet3/wallets/test2/database ErrorFile=/home/hebasto/.bitcoin/testnet3/wallets/test2/db.log
2020-04-27T21:26:04Z [] [test2] Wallet File Version = 169900
2020-04-27T21:26:04Z [] [test2] Keys: 2001 plaintext, 0 encrypted, 2001 w/ metadata, 2001 total. Unknown wallet records: 0
2020-04-27T21:26:04Z [] [test2] Wallet completed loading in              26ms
2020-04-27T21:26:04Z [] GUI: TransactionTablePriv::refreshWallet
2020-04-27T21:26:04Z [] [test2] setKeyPool.size() = 2000
2020-04-27T21:26:04Z [] [test2] mapWallet.size() = 0
2020-04-27T21:26:04Z [] [test2] m_address_book.size() = 0
```

With **this PR**:
- thread list from OS:
![Screenshot from 2020-04-28 00-21-40](https://user-images.githubusercontent.com/32963518/80425527-7a13d180-88ec-11ea-8a34-dfc774bb1c75.png)
- log excerpt:
```
2020-04-27T21:21:25Z [qt-init] GUI: initialize : Running initialization in thread
...
2020-04-27T21:23:08Z [qt-walletctrl] Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2020-04-27T21:23:08Z [qt-walletctrl] Using wallet /home/hebasto/.bitcoin/testnet3/wallets/test2
2020-04-27T21:23:08Z [qt-walletctrl] BerkeleyEnvironment::Open: LogDir=/home/hebasto/.bitcoin/testnet3/wallets/test2/database ErrorFile=/home/hebasto/.bitcoin/testnet3/wallets/test2/db.log
2020-04-27T21:23:08Z [qt-walletctrl] init message: Loading wallet...
2020-04-27T21:23:08Z [qt-walletctrl] BerkeleyEnvironment::Open: LogDir=/home/hebasto/.bitcoin/testnet3/wallets/test2/database ErrorFile=/home/hebasto/.bitcoin/testnet3/wallets/test2/db.log
2020-04-27T21:23:08Z [qt-walletctrl] [test2] Wallet File Version = 169900
2020-04-27T21:23:08Z [qt-walletctrl] [test2] Keys: 2001 plaintext, 0 encrypted, 2001 w/ metadata, 2001 total. Unknown wallet records: 0
2020-04-27T21:23:08Z [qt-walletctrl] [test2] Wallet completed loading in              37ms
2020-04-27T21:23:08Z [qt-walletctrl] init message: Rescanning...
2020-04-27T21:23:08Z [qt-walletctrl] [test2] Rescanning last 112924 blocks (from block 1609206)...
2020-04-27T21:23:08Z [qt-walletctrl] [test2] Rescan started from block 000000000000003761c81f7efbd8cebf217f39d353ec1ac59c624ac2dddfc2a8...
2020-04-27T21:23:22Z [qt-walletctrl] [test2] Rescan completed in           14157ms
2020-04-27T21:23:22Z [qt-walletctrl] GUI: TransactionTablePriv::refreshWallet
2020-04-27T21:23:22Z [qt-walletctrl] [test2] setKeyPool.size() = 2000
2020-04-27T21:23:22Z [qt-walletctrl] [test2] mapWallet.size() = 0
2020-04-27T21:23:22Z [qt-walletctrl] [test2] m_address_book.size() = 0
```

hebasto committed:
qt: Name RPCConsole executor QThread

hebasto committed:
qt: Name WalletController worker QThread

hebasto committed:
qt: Name ClientModel timer QThread

hebasto committed:
qt: Rename qt-init thread before logging start

MarcoFalke commented on src/txdb.cpp:
```diff
@@ -248,6 +249,9 @@ bool CBlockTreeDB::ReadFlag(const std::string &name, bool &fValue) {
 bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams, std::function<CBlockIndex*(const uint256&)> insertBlockIndex)
 {
+    std::string thread_name = util::ThreadGetInternalName();
+    util::ThreadRename("leveldb");
```
Why is this needed? Also, calling this leveldb is a layer-violation.

hebasto replied:
leveldb spawn a thread which inherits current thread name: "bitcoind" or "qt-init". Therefore, on master the OS reports about two "bitcoind" threads.

MarcoFalke replied:
Oh, I see. Maybe add a comment?

hebasto replied:
Mind suggesting a good wording?

laanwj replied:
To be honest, I think it's better to leave this out than do an ugly layer violation like this.
At the least this would need to be documented very well, just having a scattered thread rename is confusing for anyone reading the code in the future.
But I wonder if there's a better way to hook leveldb's thread creation and rename the threads.

hebasto replied:
Is it ok to just drop a commit with this change for better solution?

MarcoFalke replied:
It seemed weird anyway to to gui changes in the same pull request with leveldb changes

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18790#issuecomment-622099051).

laanwj commented:
Please keep to the convention of starting thread names with `b-`.

hebasto commented:
> Please keep to the convention of starting thread names with `b-`.

It is responsibility of the `ThreadRename()` function: https://github.com/bitcoin/bitcoin/blob/36c0abd8f61ba859d53b1e59014720282c97c143/src/util/threadnames.cpp#L57-L61

:)

laanwj commented:
> It is responsibility of the ThreadRename() function: 

Thanks for clearing that up.

hebasto commented:
Updated da42cd4ac1e1574a45578b04aab6e2a39653fc0e -> 64815f1f0da3431f2c2f880aeef4d57c3acd8aa5 ([pr18790.01](https://github.com/hebasto/bitcoin/commits/pr18790.01) -> [pr18790.02](https://github.com/hebasto/bitcoin/commits/pr18790.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18790.01..pr18790.02)):

- dropped _"Name leveldb thread with its own name"_ commit

jb55 commented:
ACK 64815f1f0da3431f2c2f880aeef4d57c3acd8aa5

brakmic commented:
ACK 64815f1f0da3431f2c2f880aeef4d57c3acd8aa5

Built, run and tested on macOS Catalina 10.15.4

![about_window](https://raw.githubusercontent.com/brakmic/bazaar/master/images/random/btc-about-window-1.png)

```shell
./src/qt/bitcoin-qt -logthreadnames -debug -printtoconsole -regtest                      (7s 63ms)  
2020-05-01T09:48:48Z Ignoring unknown configuration value regtest.natmap
2020-05-01T09:48:48Z [main] Bitcoin Core version pr18790.02 (release build)
2020-05-01T09:48:48Z [main] Qt 5.14.1 (dynamic), plugin=cocoa (dynamic)
2020-05-01T09:48:48Z [main] System: macOS 10.15, x86_64-little_endian-lp64
2020-05-01T09:48:48Z [main] Screen: Color LCD 1440x900, pixel ratio=1.0
2020-05-01T09:48:48Z [main] GUI: Populating font family aliases took 197 ms. Replace uses of missing font family ".AppleSystemUIFont,13,-1,5,50,0,0,0,0,0" with one that exists to avoid this cost. 
2020-05-01T09:48:48Z [main] Validating signatures for all blocks.
2020-05-01T09:48:48Z [main] Setting nMinimumChainWork=0000000000000000000000000000000000000000000000000000000000000000
2020-05-01T09:48:48Z [main] Using the 'sse4(1way),sse41(4way),avx2(8way)' SHA256 implementation
2020-05-01T09:48:48Z [main] Using RdSeed as additional entropy source
2020-05-01T09:48:48Z [main] Using RdRand as an additional entropy source
2020-05-01T09:48:48Z [main] GUI: requestInitialize : Requesting initialize
2020-05-01T09:48:48Z [qt-init] GUI: initialize : Running initialization in thread
2020-05-01T09:48:48Z [qt-init] Default data directory /Users/brakmic/Library/Application Support/Bitcoin
2020-05-01T09:48:48Z [qt-init] Using data directory /Users/brakmic/Library/Application Support/Bitcoin/regtest
2020-05-01T09:48:48Z [qt-init] Config file: /Users/brakmic/Library/Application Support/Bitcoin/bitcoin.conf
2020-05-01T09:48:48Z [qt-init] Config file arg: regtest="0"
2020-05-01T09:48:48Z [qt-init] Config file arg: testnet="0"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] daemon="1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] deprecatedrpc="signrawtransaction"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] listen="1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] port="8444"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] prune="550"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rest="1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rpcallowip="127.0.0.1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rpcbind=****
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rpcpassword=****
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rpcport="8445"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] rpcuser=****
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] server="1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] shrinkdebugfile="1"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] zmqpubhashblock="tcp://127.0.0.1:58335"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] zmqpubhashtx="tcp://127.0.0.1:58334"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] zmqpubrawblock="tcp://127.0.0.1:58332"
2020-05-01T09:48:48Z [qt-init] Config file arg: [main] zmqpubrawtx="tcp://127.0.0.1:58333"
```

hebasto commented:
@brakmic 
Thank you for reviewing. Mind particularly testing the changed threads: `qt-clientmodl`, `qt-rpcconsole`, and `qt-walletctrl`?

brakmic commented:
> @brakmic
> Thank you for reviewing. Mind particularly testing the changed threads: `qt-clientmodl`, `qt-rpcconsole`, and `qt-walletctrl`?

Yes, sure. 

I was able to get `qt-rpcconsole` and `qt-walletctrl` but no matter what I do and whichever window I open, there is no `qt-clientmodl` in sight. 

![qt-rpcconsole](https://raw.githubusercontent.com/brakmic/bazaar/master/images/random/qt-rpcconsole.png)

![qt-walletctrl](https://raw.githubusercontent.com/brakmic/bazaar/master/images/random/qt-walletctrl.png)

hebasto commented:
@brakmic 
> ... but no matter what I do and whichever window I open, there is no `qt-clientmodl` in sight.

It's because there is no logging in the `qt-clientmodl` thread. But `b-qt-clientmodl` should be showed by the OS via tools similar to `ps`, `top` etc.

brakmic commented:
> @brakmic
> 
> > ... but no matter what I do and whichever window I open, there is no `qt-clientmodl` in sight.
> 
> It's because there is no logging in the `qt-clientmodl` thread. But `b-qt-clientmodl` should be showed by the OS via tools similar to `ps`, `top` etc.

Ok. 

I took a sample from the running process and here's the `b-qt-clientmodl`thread.

![qt-clientmodel](https://raw.githubusercontent.com/brakmic/bazaar/master/images/random/b-qt-clientmodl.png)

hebasto commented:
Rebased 64815f1f0da3431f2c2f880aeef4d57c3acd8aa5 -> 6d8747dd13dc8d7c4868929041357ad4dda5aad8 ([pr18790.02](https://github.com/hebasto/bitcoin/commits/pr18790.02) -> [pr18790.03](https://github.com/hebasto/bitcoin/commits/pr18790.03)) due to the conflict with #18699.

hebasto commented:
Rebased 6d8747dd13dc8d7c4868929041357ad4dda5aad8 -> f979110c851372b8702c3ae99e21f2e578640e4b ([pr18790.03](https://github.com/hebasto/bitcoin/commits/pr18790.03) -> [pr18790.04](https://github.com/hebasto/bitcoin/commits/pr18790.04)) due to the conflict with #18152.

promag reviewed (commented):
Concept ACK, build and dumped all thread list with lld, looks good.

From https://doc.qt.io/qt-5/qthread.html

> To choose the name that your thread will be given (as identified by the command ps -L on Linux, for example), you can call setObjectName() before starting the thread. If you don't call setObjectName(), the name given to your thread will be the class name of the runtime type of your thread object (for example, "RenderThread" in the case of the Mandelbrot Example, as that is the name of the QThread subclass). Note that this is currently not available with release builds on Windows.

Is it because of the above windows issue that you don't use `setObjectName`? Not sure if that even maters to us.

hebasto commented:
@promag 
> Is it because of the above windows issue that you don't use `setObjectName`? Not sure if that even maters to us.

Using `setObjectName()` does not work for our logging with thread names.

jonatack commented:
ACK f979110c851372b8702c3ae99e21f2e578640e4b

jonatack reviewed (commented):
ACK f979110

Sjors commented:
Concept ACK

hebasto commented:
Rebased f979110c851372b8702c3ae99e21f2e578640e4b -> ead771bf6fc7a4b96a03d4938796c88657c69ba6 ([pr18790.04](https://github.com/hebasto/bitcoin/commits/pr18790.04) -> [pr18790.05](https://github.com/hebasto/bitcoin/commits/pr18790.05)) due to the conflict with #19011.

fanquake commented:
Planning on reviewing (shortly) and merging this PR in this repo. @promag @jb55 @jonatack @Sjors want to take another look here?

Sjors commented:
tACK ead771bf6fc7a4b96a03d4938796c88657c69ba6
